### PR TITLE
pc/storage/kdtree: add minimum distance option

### PR DESCRIPTION
**Pull request by at-wat**
Support approximated search.

```
BenchmarkKDTree_Nearest/minDistSq=0.00/100000points/KDTree-32  822694  1420 ns/op   153 B/op  6 allocs/op
BenchmarkKDTree_Nearest/minDistSq=0.01/100000points/KDTree-32  956205  1256 ns/op   134 B/op  5 allocs/op
BenchmarkKDTree_Nearest/minDistSq=0.10/100000points/KDTree-32 4374184   275.3 ns/op  24 B/op  1 allocs/op
```

### CODE REVIEW

File: pc/storage/kdtree/kdtree_test.go [hunk 1]
```diff
@@ -122,601 +122,608 @@ func kdtreeDeepExpectEqual(t *testing.T, a, b *KDTree) bool {
 }
 
 func TestKDtree(t *testing.T) {
-	it := createTestPointCloud(t)
-	kdt := New(it)
+	for _, minDist := range []float32{0, 0.001} {
```
Comment (f-fl0) [inline pc/storage/kdtree/kdtree_test.go:125]: [optional] Some of the tests do not seem to be affected by `minDistSq` (e.q. tree construction, range, delete). Maybe they could be defined outside of the for loop.

File: pc/storage/kdtree/kdtree.go [hunk 2]
```diff
@@ -16,6 +16,10 @@ type KDTree struct {
 	root *node
 
 	poolNodeArray *sync.Pool
+
+	// Squared distance to cut the search branch.
+	// If this value is larger than zero, Nearest will be approximated search.
+	MinDistSq float32
```
Comment (f-fl0) [inline pc/storage/kdtree/kdtree.go:22]: Do you think it is better to make it a property of the tree more than an argument of the `Nearest` method? I guess one can also set `MinDistSq` to 0 if the original behavior of `Nearest` is needed without rebuilding the tree.

**Review by f-fl0 — COMMENTED** [review]
Just left a few comments.

File: pc/storage/kdtree/kdtree.go (hunk 2 above)
Comment (at-wat) [reply pc/storage/kdtree/kdtree.go:22]: I thought it, but it makes `MinDistSq` general parameter of `storage.Search` interface.
`MinDistSq` is kd-tree (some other may have same) specific parameter, so I don't want to put it to general `storage.Search`.

Comment (f-fl0) [reply pc/storage/kdtree/kdtree.go:22]: It makes sense.

Comment (at-wat) [reply pc/storage/kdtree/kdtree.go:22]: Added `KDTree.With(...KDTreeOption)` method to create shallow copy of the tree with different options.
https://github.com/seqsense/pcgol/pull/29/files#diff-406022fe6d5d2a5bc87edb91256ccbb800bd4f4eff7efcd36dd3b3a0eb3741dbR157-R161

**Pull request merged**